modules/core/src/main/resources/reference.conf

```diff
@@ -21,8 +21,9 @@
   "batching": {
     "maxBytes": 16000000
     "maxDelay": "1 second"
-    "uploadConcurrency":  3
+    "uploadParallelismFactor":  2.5
   }
+  "cpuParallelismFactor": 0.75
```

Comment:
> We set uploadParallelismFactor much higher than cpuParallelismFactor to be 100% sure that the uploading will never be a bottleneck...

I wouldn't compare the two values directly like that, because they control very different things.  We fan-out to parse/transform a batch in parallel across many cpu, and then join the result to produce a single batch.  Next, we upload the batch, and that is a long slow step.  We can run many many uploads in parallel so it doesn't really matter that uploading is slow.  Whereas for cpu-intensive steps we are limited by the number of cpu available  The cpu-intensive steps are much faster than the uploading steps.

So... think of them as two independently tunable parameters.  Both parameters should be just large enough to keep the cpu always busy, and not too much larger.

> ...but setting them roughly the same should work just fine.

If you decrease `uploadParallelismFactor` to be `0.75` then uploading will become the bottleneck.

On the other hand if you increase `cpuParallelismFactor` to be `2.5.` then it _probably_ works ok... except that it doesn't really make sense for cpu parallelism to be larger than then the number of cpu available.

> Even though 0.75 * number of CPUs batches can be transformed in parallel, all the CPUs are used to process them.

I'll explain using an example where there are 4 CPU available, so cpu parallelism is equal to 3.  Meaning it will parse/transform 3 batches at once).

From a cpu point of view:  Every cpu is always doing _something_.  Which might be parsing/transforming some events, or it might be anything else.... like fetching from kinesis or check-pointing or uploading or sending metrics or..... anything.

From a batch point of view:  The 3 parallel batches might each get parsed/transformed on a different cpu.  On the other hand, maybe 3 of the cpu are already very busy doing the _other_ things listed above.  In which case, our three batches might all get parsed/transformed on the same cpu.  That all depends on how cats-effect schedules the fibers across different threads, and how linux schedules the threads across different cpus

You cannot directly control what happens on which cpu.  But what we can do is provide enough parallelism in our code, so that the runtime environment (cats-effect + JVM + the kernel's scheduler) can keep the cpu fully busy all the time.

Reply:
That all makes perfect sense, thanks a lot for the explanation :+1: